pkg/controllers/metrics/node/counter.go

```diff
@@ -0,0 +1,243 @@
+			})))
+		}
+
+		for _, instanceType := range instanceTypeValues {
+			// 4. Publish the count of all nodes with `instanceType`, associated with `provisioner`, in `zone`, and reported as "ready"
+			nodeLabels := client.MatchingLabels{
+				nodeLabelInstanceType: instanceType,
+				nodeLabelProvisioner:  provisioner,
+				nodeLabelZone:         zone,
+			}
+			errors = append(errors, consumeNodesWith(nodeLabels, filterReadyNodes(func(readyNodes []corev1.Node) error {
+				return publishCount(readyNodeCountByInstancetypeProvisionerZone, metricLabelsFromNodeLabels(nodeLabels), len(readyNodes))
+			})))
+		}
+
+		for _, os := range osValues {
+			// 5. Publish the count of all nodes with `os`, associated with `provisioner`, in `zone`, and reported as "ready".
+			nodeLabels := client.MatchingLabels{
+				nodeLabelOs:          os,
+				nodeLabelProvisioner: provisioner,
+				nodeLabelZone:        zone,
+			}
+			errors = append(errors, consumeNodesWith(nodeLabels, filterReadyNodes(func(readyNodes []corev1.Node) error {
+				return publishCount(readyNodeCountByOsProvisionerZone, metricLabelsFromNodeLabels(nodeLabels), len(readyNodes))
+			})))
+		}
+	}
+
+	// Combine will filter out `nil` values; if no errors remain then it will return `nil`.
+	return multierr.Combine(errors...)
+}
+
+// filterReadyNodes returns a new function that will filter "ready" nodes to pass on
+// to `consume`, and returns the result.
+func filterReadyNodes(consume nodeListConsumerFunc) nodeListConsumerFunc {
+	return func(nodes []corev1.Node) error {
+		readyNodes := make([]corev1.Node, 0, len(nodes))
+		for _, node := range nodes {
+			for _, condition := range node.Status.Conditions {
+				if condition.Type == nodeConditionTypeReady && strings.ToLower(string(condition.Status)) == "true" {
+					readyNodes = append(readyNodes, node)
+				}
+			}
+		}
+		return consume(readyNodes)
+	}
+}
+
+func metricLabelsFromNodeLabels(nodeLabels client.MatchingLabels) (metricLabels prometheus.Labels) {
+	metricLabels = prometheus.Labels{}
+	// Exclude node label values that not present or are empty strings.
+	if arch := nodeLabels[nodeLabelArch]; arch != "" {
+		metricLabels[metricLabelArch] = arch
+	}
+	if instanceType := nodeLabels[nodeLabelInstanceType]; instanceType != "" {
+		metricLabels[metricLabelInstanceType] = instanceType
+	}
+	if os := nodeLabels[nodeLabelOs]; os != "" {
+		metricLabels[metricLabelOs] = os
+	}
+	if provisioner := nodeLabels[nodeLabelProvisioner]; provisioner != "" {
+		metricLabels[metricLabelProvisioner] = provisioner
+	}
+	if zone := nodeLabels[nodeLabelZone]; zone != "" {
+		metricLabels[metricLabelZone] = zone
+	}
+	return
+}
+
+func publishCount(gaugeVec *prometheus.GaugeVec, labels prometheus.Labels, count int) (err error) {
+	var gauge prometheus.Gauge
+	gauge, err = gaugeVec.GetMetricWith(labels)
+	if err != nil {
+		return
+	}
+	gauge.Set(float64(count))
+	return
```

Comment:
This can be simplified to:

```go
	gauge, err := gaugeVec.GetMetricWith(labels)
	if err == nil {
	    gauge.Set(float64(count))
        }
	return err
```

Also, the return value would go from `(err error)` -> `error`.